task_interpreter_ipf_client/models/client.py

```diff
@@ -234,37 +234,66 @@ def populate_all_data(self, silent=False):
         ok = True
-        for method, name, fields in (
-                (self.get_tolksprak, 'res.interpreter.language',
-                 (('name', 'namn'), ('code', 'id'))),
-                (self.get_kon, 'res.interpreter.gender_preference',
-                 (('name', 'namn'), ('code', 'id'))),
-                (self.get_tolktyp, 'res.interpreter.type',
-                 (('name', 'namn'), ('code', 'id'))),
-                (self.get_distanstolktyp, 'res.interpreter.remote_type',
-                 (('name', 'namn'), ('code', 'id')))
-        ):
-
-            if not self._populate_data(method, name, fields):
+        for method, key, name, field_spec in (
+                (self.get_tolksprak, 'code', 'res.interpreter.language',
+                 {'name': 'namn', 'code': 'id'}),
+                (self.get_kon, 'code', 'res.interpreter.gender_preference',
+                 {'name': 'namn', 'code': 'id'}),
+                (self.get_tolktyp, 'code', 'res.interpreter.type',
+                 {'name': 'namn', 'code': 'id'}),
+                (self.get_distanstolktyp, 'code', 'res.interpreter.remote_type',
+                 {'name': 'namn', 'code': 'id'})):
+            if not self._populate_data(method, key, name, field_spec):
                 msg = f'Failed to populate data for {name}'
-                _logger.warn(msg)
+                _logger.warning(msg)
                 if not silent:
                     raise Warning(msg)
                 ok = False
         return ok
 
     @api.model
-    def _populate_data(self, method, model_name, fields):
+    def _populate_data(self, method, key, model_name, field_spec):
+        """
+        Populates interpreter res storages with data.
+
+        Removes entries that does not exists anymore, updates changed
+        elements and creates new if they dont exists in db yet.
+        """
         result = method()
         if result.status_code not in (200, 201):
-            _logger.warn(f'Failed to populate {model_name} with code: '
-                         f'{result.status_code} and message: {result.text}')
+            _logger.warning(f'Failed to populate {model_name} with code: '
+                            f'{result.status_code} and message: {result.text}')
             return False
-        self.env[model_name].search([]).unlink()
-        for entry in json.loads(result.text):
-            self.env[model_name].create(
-                {field_name: entry[result_name] for field_name, result_name in
-                 fields})
+        mod = self.env[model_name]
+        res = json.loads(result.text)
+        search_term = [(key, 'in', [r[field_spec[key]] for r in res])]
+        # Removing entries that have been removed.
+        in_db = mod.search([])
+        matching = mod.search(search_term)
+        _logger.debug(f'Unlinking: {in_db - matching}')
+        (in_db - matching).unlink()
+
+        unchanged = mod.search([(field,
```

Comment:
Sökningen matchar inte alla fältvärdena till samma objekt. Det finns en risk att få falska träffar här. Detta kan förbättras med &- och |-operatorer i sökdomänen så att vi bara matchar exakt de .
```
domain = []
for r in res:
    # Addera OR mellan objekten
    if domain:
        domain.insert(0, '|')
    # Addera AND mellan aktuellt objekts fält
    for x in range(len(field_spec) - 1):
        domain.append('&')
    for field in field_spec:
        domain.append((field, '=',  r[field_spec[field]]))
```
Det borde resultera in en sökdomän som ser ut ungefär så här (med lämpligt antal | och &):
```
[
    '|',
        '&',
            (field1, '=', obj1_field1),
            (field2, '=', obj1_field2),
        '&',
            (field1, '=', obj2_field1),
            (field2, '=', obj2_field2),
]
```
Om vi bara vill ha ut värdet på ett fält så är det dessutom ett bra läge att använda search_read för lite prestanda-optimering. Inte värt så många CPU-cykler just här, men när man ändå håller på.
```
unchanged = [x[key] for x in search_read(domain, [key])]
```